File: AddFraction.vb
Public Class AddFraction

    Protected numerators() As Integer
    Protected denominators() As Integer
    Protected new_numerators() As Integer

    Protected lcm As Integer
    Protected answer As Integer

    ' Simulate a constructor
    Public Sub _init_(fractions As Dictionary(Of String, Integer()))
        numerators = fractions.Item("numerators")
        denominators = fractions.Item("denominators")
        ReDim new_numerators(numerators.Count - 1)
        answer = 0
    End Sub

    ' transforms fractions so they all have same denominator
    ' takes a dictionary Of a list Of numerators And denominators
    '
    ' returns a dictionary Of the New numerators And New denominator(LCM)
    Protected Sub canonizeFraction()
        ' STEP 1:
        Dim l_c_m As New LCM
        l_c_m._init_(denominators)
        lcm = l_c_m.getLCM()

        ' STEP 2:
        ' make numerators vary(ratio) With lcm
        For index = 0 To denominators.Count - 1
            new_numerators(index) = lcm / denominators(index) * numerators(index)
        Next index

    End Sub

    ' returns a dictionary Of the
    ' New numerator And New denominator(LCM)
    Public Function doAdd() As Dictionary(Of String, Integer)
        canonizeFraction()

        ' STEP 3:
        ' add all transformed numerators
        For Each num In new_numerators
            answer += num
        Next

        Dim send_back As New Dictionary(Of String, Integer)
        send_back.Add("numerator", answer)
        send_back.Add("denominator", LCM)
        Return send_back

    End Function

End Class

File: AddFraction.py
# A class
class PlusFraction:

    # Simulate a constructor
    def __init__(self, fractions):
        
        self.numerators   = fractions['numerators']
        self.denominators = fractions['denominators']

        self.answer = 0
        self.new_numerators = []


    # transforms fractions so they all have same denominator
    # takes a dictionary of a list of numerators and denominators
    #
    # returns a dictionary of the new numerators and new denominator(LCM)
    def canonizeFraction(self):
		# STEP 1:
        from LCM import findLCM
        self.l_c_m = findLCM(self.denominators)
        self.lcm = self.l_c_m.getLCM()

		# STEP 2:
        # make numerators vary(ratio) with lcm
        for count in range(len(self.denominators)):
            self.new_numerators.append(int(self.lcm / self.denominators[count] * self.numerators[count]))

        return None

    # returns a dictionary of the
    # new numerator and new denominator(LCM)
    def doAdd(self):
        self.canonizeFraction()
        
		# STEP 3:
		# add all transformed numerators
        for num in self.new_numerators: self.answer += num
        return {'numerator':self.answer, 'denominator':self.lcm}

